### PR TITLE
[ATOM-16021] Initial continuous capture support 

### PR DESCRIPTION
**Pull request by jcbhl**
### Please [start a build](https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-2624/) and request sig-presentation!

This PR implements capturing multiple frames of CPU profiling data and saving it to disk so that it can be analyzed offline (which will be implemented in the visualizer in a future PR). Now, the ProfilingCaptureSystemComponent offers endpoints for capturing one frame of profiling data (with a blocking IO call) in addition to an arbitrary amount of frames (using an IO thread). 

### Demo
https://user-images.githubusercontent.com/64656371/127545500-06bed756-0d15-48ff-8966-6ddf402af690.mp4
image
Resulting file size on disk was 70 MB from ~3000 frames of data.

### CODE REVIEW

File: Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp [hunk 1]
```diff
@@ -590,6 +600,50 @@ namespace AZ
...
+            }
+
+            const bool captureEnded = AZ::RHI::CpuProfiler::Get()->EndContinuousCapture(m_lastContinuousCapture);
+            if (!captureEnded)
+            {
+                AZ_TracePrintf("ProfilingCaptureSystemComponent", "Could not end the continuous capture, is one in progress?\n");
+                return false;
+            }
+
+            // cpuProfilingData could be 1GB+ once saved, so use an IO thread to write it to disk.
```
Comment (jcbhl) [inline Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp:632]: Note: please be super critical of my use of threading here, I haven't used threading primitives before so my use is likely incorrect 😁

File: Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.h [hunk 2]
```diff
@@ -84,6 +87,13 @@ namespace AZ
             DelayedQueryCaptureHelper m_pipelineStatisticsCapture;
             DelayedQueryCaptureHelper m_cpuProfilingStatisticsCapture;
             DelayedQueryCaptureHelper m_benchmarkMetadataCapture;
+
+            // Stores the last continuous capture data so that the IO thread can access it by reference. 
+            // Ensure that this is not cleared until m_cpuDataSerializationInProgress is false. 
+            AZStd::deque<RHI::CpuProfiler::TimeRegionMap> m_lastContinuousCapture;
```
Comment (jcbhl) [inline Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.h:93]: Note: I looked into the lifetime of `ProfilingCaptureSystemComponent` and it appears that the component is only destructed when the renderer is destructed (see `CommonModule.cpp`) so this deque should always be alive while the IO thread is working.

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CpuProfilerImpl.h [hunk 3]
```diff
@@ -134,6 +139,12 @@ namespace AZ
             AZStd::shared_mutex m_shutdownMutex;
 
             bool m_initialized = false;
+
+            bool m_continuousCaptureInProgress = false;
+
+            // Stores multiple frames of profiling data, size is controlled by MaxFramesToSave. Flushed when EndContinuousCapture is called.
+            // Deque so that we can have fast append of new data + removal of old profiling data.
+            AZStd::deque<TimeRegionMap> m_continuousCaptureData;
```
Comment (jcbhl) [inline Gems/Atom/RHI/Code/Include/Atom/RHI/CpuProfilerImpl.h:147]: Question: thoughts about using deque here over a vector? I chose it so that we could have O(1) removal of old profiling data if we are at `MaxFramesToSave` but it may perform worse than just a plain old vector due to cache effects.

**Review by wintermute-motherbrain — APPROVED** [review]

File: Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp [hunk 4]
```diff
@@ -590,6 +600,50 @@ namespace AZ
...
+            m_cpuDataSerializationInProgress.store(true);
+            auto thread = AZStd::thread(
+                [&data = m_lastContinuousCapture, filePath = AZStd::string(outputFilePath), &flag = m_cpuDataSerializationInProgress]()
+                {
+                    SerializeCpuProfilingData(data, filePath, true);
+                    flag.store(false);
+                });
+
+            thread.detach();
```
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp:641]: It looks like you're joining this thread later, so you shouldn't detach the thread at this point I don't think

File: Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp [hunk 5]
```diff
@@ -590,6 +600,50 @@ namespace AZ
...
+            const bool captureEnded = AZ::RHI::CpuProfiler::Get()->EndContinuousCapture(m_lastContinuousCapture);
+            if (!captureEnded)
+            {
+                AZ_TracePrintf("ProfilingCaptureSystemComponent", "Could not end the continuous capture, is one in progress?\n");
+                return false;
+            }
+
+            // cpuProfilingData could be 1GB+ once saved, so use an IO thread to write it to disk.
+            m_cpuDataSerializationInProgress.store(true);
+            auto thread = AZStd::thread(
```
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp:634]: Do we not have a thread pool for enqueuing async work of this nature as opposed to spinning up a dedicated thread, or is the idea to keep work off the async thread pool for some reason?

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CpuProfilerImpl.h (hunk 3 above)
Comment (jeremyong-az) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/CpuProfilerImpl.h:147]: You could use a circular or ring buffer to get the benefits of both O(1) removal/insertion in a linear range of memory like a vector.

File: Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp (hunk 5 above)
Comment (jcbhl) [reply Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp]: No specific reason for using an owned thread here - moved to `AZ::Job` which should avoid some of the overhead of spawning a new thread!

File: Gems/Atom/RHI/Code/Include/Atom/RHI/CpuProfilerImpl.h (hunk 3 above)
Comment (jcbhl) [reply Gems/Atom/RHI/Code/Include/Atom/RHI/CpuProfilerImpl.h]: That's exactly what I was looking for, thanks! Moved to `AZStd::ring_buffer`

**Review by hershey5045 — APPROVED** [review]

File: Gems/Atom/RHI/Code/Source/RHI/CpuProfilerImpl.cpp [hunk 6]
```diff
@@ -141,12 +144,47 @@ namespace AZ
             return m_timeRegionMap;
         }
 
+        bool CpuProfilerImpl::BeginContinuousCapture()
+        {
+            if (m_continuousCaptureInProgress)
```
Comment (jeremyong-az) [inline Gems/Atom/RHI/Code/Source/RHI/CpuProfilerImpl.cpp:149]: After loading the value in this atomic, nothing stops a different thread from entering the same place and setting `m_continuousCaptureInProgress` to true on line 156, before _this thread_ gets to the same line. While this is technically benign because the atomic will stay true in both cases, you might have two threads competing to deactivate the capture later because they both believe they started the capture.

The pattern to employ here is the atomic exchange, where you do something like

```cpp
if (m_continuousCaptureInProgress.compare_exchange_strong(false, true))
{
    // We atomically transitioned the value of m_continuousCaptureInProgress from false to true, so are responsible for the capture and can start it
}
else
{
    return false;
}
```

**Comment by jcbhl** [comment]
Thanks again for pointing out the issues here @jeremyong-az. I went back on unifying the state of the CpuProfiler and the ProfilingCaptureSystemComponent - I want to separate the end use of the data from the internal implementation, as some devs might want to do in-memory work with the data that doesn't require writing to disk (ex. calculating aggregate statistics). 


I switched to using `compare_exchange_strong` for the atomics like you suggested, and also added a mutex around `EndContinuousCapture()`. I realized that multiple threads could enter that function and move out of `m_continuousCaptureData` more than once, which I believe is UB. Also added [[nodiscard]] to better indicate that the functions may fail, ideally this would be an `AZ::Outcome` but I wanted to keep the changes small here.

File: Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp [hunk 7]
```diff
@@ -603,55 +603,49 @@ namespace AZ
...
-            
-            // If the thread object already exists (ex. we have already serialized data), join. This will not block since
-            // m_cpuDataSerializationInProgress is false, meaning the IO thread has already completed execution.
-            // TODO Use a reusable thread implementation over repeated creation + destruction of threads [ATOM-16214]
-            if (m_cpuDataSerializationThread.joinable())
-            {
-                m_cpuDataSerializationThread.join();
-            }
+                auto thread = AZStd::thread(threadIoFunction);
+                m_cpuDataSerializationThread = AZStd::move(thread);
```
Comment (jeremyong-az) [inline Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp:640]: Random nit, you can just do `m_cpuDataSerializationThread = AZStd::thread{threadIoFunction}` here.

**Review by jeremyong-az — APPROVED** [review]
I will say, I'm not 100% convinced myself that this is necessarily thread safe from just a visual code inspection myself, as there are still a number of complex interactions. I will approve but recommend independent testing or simplification in cases where the atomic + mutex interaction isn't strictly necessary for performance.

**Comment by jcbhl** [comment]
Definitely agree - this is pretty messy at the moment. I'll put up a jira for adding automated tests + cleaning this up. Thanks!
